File: packages/music-player/truepeak-core.bench/src/buckets.rs
//! The bucket-first composite: per-provenance coverage and margin, with the FLAC
//! bones-guided probe.
//!
//! Buckets come from embedded tags only (codec, store IDs, iTunNORM, youtube provenance;
//! never path text). Their zoom tails diverge hard: FLAC's frame sizes track signal level
//! (lossless bits follow residual entropy), so a bones-guided probe keeps FLAC accurate at
//! a fraction of the decode cost, and the freed seconds buy the risky untagged-lossy
//! bucket more coverage. Perceptual codecs' byte profiles were measured useless for this
//! (bits follow busyness, not height), so lossy buckets use the plain frontier zoom.

/// Imports the corpus track record.
use crate::corpus::Track;
/// Imports the max-heap driving the frontier expansion.
use std::collections::BinaryHeap;
/// Imports the tag and profile join maps.
use std::collections::HashMap;
/// Imports the line-by-line side-file readers.
use std::fs::File;
/// Imports buffered reading so the side files stream.
use std::io::{BufRead, BufReader};
/// Imports the borrowed path type for side-file locations.
use std::path::Path;

/// Imports `anyhow` helpers for application-level error returns.
use anyhow::{Context, Result};
/// Imports serde's derive so side-file rows parse straight from each JSON line.
use serde::Deserialize;
/// Imports the shared dB conversion, the policy, and the engine's bucket types.
use truepeak_core::{BucketProbe, Policy, TrackProvenance, peak_dbtp};

/// One row of the embedded-tag sweep (analysis/tags-sweep.mjs output).
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct TagRow {
    /// Track path, the join key against the corpus.
    path: String,
    /// Audio codec name from ffprobe (`flac`, `opus`, `aac`, `mp3`).
    codec: Option<String>,
    /// Whether store identifiers (ISRC/UPC/content ids) are embedded.
    #[serde(default)]
    has_store_ids: bool,
    /// Whether an iTunNORM tag is embedded.
    #[serde(default)]
    has_itun_norm: bool,
    /// Whether a youtube provenance tag (purl or youtube comment) is embedded.
    #[serde(default)]
    has_purl: bool,
}

/// One row of the FLAC frame-size profile file (analysis/flac-bones.mjs output).
#[derive(Deserialize)]
struct ProfileRow {
    /// Track path, the join key against the corpus.
    path: String,
    /// Compressed bytes per 0.1 s slot from the CRC-verified frame walk.
    bytes: Vec<u64>,
}

/// The provenance bucket a track probes under.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum Bucket {
    /// Lossless FLAC: thin tail, bones-guided probe.
    Flac,
    /// Store-tagged lossy (ISRC/UPC/iTunNORM): mastered releases.
    Store,
    /// Youtube-provenance lossy: loudness-normalized sources.
    Purl,
    /// Untagged lossy: the risk bucket that gets the freed coverage.
    Bare,
}

/// The probe dial the SHIPPED policy assigns this bucket.
///
/// What: maps the bench bucket to the engine's provenance selection, so the bench
/// evaluates exactly the table `truepeak_core::default_policy` ships. Why: one source
/// of truth; the bench validates the engine's dials instead of carrying its own copy.
fn bucket_probe(policy: &Policy, bucket: Bucket, bones_present: bool) -> BucketProbe {
    let provenance = match bucket {
        Bucket::Flac => TrackProvenance { lossless: true, ..TrackProvenance::unknown() },
        Bucket::Store => TrackProvenance { store_tagged: true, ..TrackProvenance::unknown() },
        Bucket::Purl => TrackProvenance { youtube_tagged: true, ..TrackProvenance::unknown() },
        Bucket::Bare => TrackProvenance::unknown(),
    };
    provenance.select(&policy.buckets, bones_present)
}

/// Read the tag sweep into a per-path bucket map; tracks missing from the sweep fall
/// back to their file extension.
///
/// What: joins tags-full.jsonl rows to buckets. Why: bucket assignment must be
/// reproducible from committed side files, never from path text heuristics.
pub fn load_buckets(path: &Path) -> Result<HashMap<String, Bucket>> {
    // Stream rows and reduce each to its bucket.
    let file = File::open(path)?;
    let reader = BufReader::new(file);
    let mut map = HashMap::new();
    for line in reader.lines() {
        let line = line?;
        if line.trim().is_empty() {
            continue;
        }
        let row: TagRow = serde_json::from_str(&line)?;
        let bucket = if row.codec.as_deref() == Some("flac") {
            Bucket::Flac
        } else if row.has_store_ids || row.has_itun_norm {
            Bucket::Store
        } else if row.has_purl {
            Bucket::Purl
        } else {
            Bucket::Bare
        };
        map.insert(row.path, bucket);
    }
    Ok(map)
}

/// The extension-fallback bucket for tracks the tag sweep missed.
fn bucket_from_extension(path: &str) -> Bucket {
    if path.ends_with(".flac") {
        Bucket::Flac
    } else {
        Bucket::Bare
    }
}

/// Read the FLAC frame-size profiles into a per-path top-slot list.
///
/// What: keeps only each profile's `top` largest byte slots (the policy's
/// `bones_top_slots`). Why: the probe needs the slot indices, not the raw profile.
pub fn load_bones(path: &Path, top: usize) -> Result<HashMap<String, Vec<usize>>> {
    // Stream rows, sort slot indices by byte count, keep the top few.
    let file = File::open(path)?;
    let reader = BufReader::new(file);
    let mut map = HashMap::new();
    for line in reader.lines() {
        let line = line?;
        if line.trim().is_empty() {
            continue;
        }
        let row: ProfileRow = serde_json::from_str(&line)?;
        let mut order: Vec<usize> = (0..row.bytes.len()).collect();
        order.sort_by(|&a, &b| row.bytes[b].cmp(&row.bytes[a]));
        order.truncate(top);
        map.insert(row.path, order);
    }
    Ok(map)
}

/// Convert a linear peak to dBTP, treating silence as a very negative level.
fn db(peak: f64) -> f64 {
    if peak <= 0.0 {
        f64::NEG_INFINITY
    } else {
        peak_dbtp(peak)
    }
}

/// The hybrid probe: optional bones seed slots (each ±1), an even pass, then frontier
/// zoom until the bin budget is spent; returns the loudest decoded bin and decoded secs.
///
/// What: the same climb as `zoom.rs`, seeded by byte-rate hot spots when a profile
/// exists. Why: for lossless codecs the byte profile points at loud passages, so the
/// climb starts on the right hills at a fraction of the coverage.
fn hybrid_probe(track: &Track, coverage: f64, even_coverage: f64, bones: Option<&Vec<usize>>) -> (f64, f64) {
    let bins = &track.bin_peaks;
    let n = bins.len();
    let budget_bins = ((coverage * n as f64).floor() as usize).max(1);
    let mut decoded = vec![false; n];
    // Non-negative f32 bit patterns order like the values; the heap pops loudest first.
    let mut heap: BinaryHeap<(u32, usize)> = BinaryHeap::new();
    let mut used = 0usize;
    let mut peak = 0.0f32;
    let decode = |index: usize,
                  decoded: &mut Vec<bool>,
                  heap: &mut BinaryHeap<(u32, usize)>,
                  used: &mut usize,
                  peak: &mut f32| {
        decoded[index] = true;
        *used += 1;
        if bins[index] > *peak {
            *peak = bins[index];
        }
        heap.push((bins[index].to_bits(), index));
    };
    if let Some(slots) = bones {
        for &slot in slots {
            let lo = slot.saturating_sub(1);
            let hi = (slot + 1).min(n.saturating_sub(1));
            for index in lo..=hi {
                if !decoded[index] && used < budget_bins {
                    decode(index, &mut decoded, &mut heap, &mut used, &mut peak);
                }
            }
        }
    }
    let count = ((even_coverage * n as f64).round() as usize).max(1);
    let span = n - 1;
    for step in 0..count {
        if used >= budget_bins {
            break;
        }
        let index = if count <= 1 {
            span / 2
        } else {
            ((step as f64 / (count - 1) as f64) * span as f64).round() as usize
        };
        if !decoded[index] {
            decode(index, &mut decoded, &mut heap, &mut used, &mut peak);
        }
    }
    while used < budget_bins {
        let Some((_, index)) = heap.pop() else {
            break;
        };
        if index > 0 && !decoded[index - 1] && used < budget_bins {
            decode(index - 1, &mut decoded, &mut heap, &mut used, &mut peak);
        }
        if index + 1 < n && !decoded[index + 1] && used < budget_bins {
            decode(index + 1, &mut decoded, &mut heap, &mut used, &mut peak);
        }
    }
    (f64::from(peak), used as f64 * track.bin_seconds)
}

/// One probed long track under the composite: levels, margin, and its bucket.
#[derive(Clone, Debug)]
pub struct BucketRow {
    /// Full-track true peak in dBTP.
    pub full_db: f64,
    /// Loudest decoded bin in dBTP.
    pub probe_db: f64,
    /// The margin its bucket assigns in dB.
    pub margin_db: f64,
    /// The bucket, for clamp breakdowns.
    pub bucket: Bucket,
}

/// Print the decided composite's report: budget, three measures, clamp breakdown.
///
/// What: loads the two side files (tag sweep, FLAC profiles), evaluates the decided
/// assignment, and prints the letter's measures. Why: the committed, reproducible
/// evaluation of the bucket-first answer.
pub fn report_buckets(
    tracks: &[Track],
    full_secs: f64,
    target_secs: f64,
    args: &[String],
) -> Result<()> {
    // The side files are the non-flag arguments after the corpus path, in order.
    let side: Vec<&String> = args.iter().skip(2).filter(|arg| !arg.starts_with("--")).collect();
    let tags_path = side
        .first()
        .context("usage: truepeak-core-bench <corpus> <tags-full.jsonl> [flac-profiles.jsonl] --buckets")?;
    let buckets = load_buckets(Path::new(tags_path))?;
    let policy = truepeak_core::default_policy();
    let bones = match side.get(1) {
        Some(profiles_path) => load_bones(Path::new(profiles_path), policy.bones_top_slots)?,
        None => HashMap::new(),
    };
    let (decoded, rows) = evaluate_buckets(tracks, &policy, &buckets, &bones);
    println!("\nbucket composite (the SHIPPED default_policy table; flac bones-guided):");
    for bucket in [Bucket::Flac, Bucket::Store, Bucket::Purl, Bucket::Bare] {
        let probe = bucket_probe(&policy, bucket, bucket == Bucket::Flac && !bones.is_empty());
        let members = rows.iter().filter(|row| row.bucket == bucket).count();
        println!(
            "  {bucket:?}: coverage={} margin={}dB long_tracks={members}",
            probe.coverage_fraction, probe.probe_margin_db
        );
    }
    println!(
        "decoded={decoded:.0}s ({:.2}% of corpus) target={target_secs:.0}s {}",
        100.0 * decoded / full_secs,
        if decoded <= target_secs { "IN BUDGET" } else { "OVER" }
    );
    // The letter's three measures over every track (shorts are exact, zero error).
    let mut clamps: HashMap<Bucket, usize> = HashMap::new();
    let mut quiet_sum = 0.0f64;
    let mut worst_quiet = 0.0f64;
    let mut worst_over = 0.0f64;
    for row in &rows {
        let necessary = (row.full_db - policy.ceiling_dbtp).max(0.0);
        let applied = (row.probe_db + row.margin_db - policy.ceiling_dbtp).max(0.0);
        quiet_sum += (applied - necessary).max(0.0);
        worst_quiet = worst_quiet.max((applied - necessary).max(0.0));
        worst_over = worst_over.max((necessary - applied).max(0.0));
        if row.full_db > policy.ceiling_dbtp
            && row.full_db - row.probe_db - row.margin_db > policy.max_too_loud_db
        {
            *clamps.entry(row.bucket).or_insert(0) += 1;
        }
    }
    let clamp_total: usize = clamps.values().sum();
    println!(
        "clamped={clamp_total} (flac={} store={} purl={} bare={}) avg_quiet={:.3}dB worst_quiet={:.2}dB worst_over={:.2}dB",
        clamps.get(&Bucket::Flac).unwrap_or(&0),
        clamps.get(&Bucket::Store).unwrap_or(&0),
        clamps.get(&Bucket::Purl).unwrap_or(&0),
        clamps.get(&Bucket::Bare).unwrap_or(&0),
        quiet_sum / tracks.len() as f64,
        worst_quiet,
        worst_over,
    );
    Ok(())
}

/// Evaluate the shipped composite over the corpus.
///
/// What: short tracks full-scan; each long track probes under the SHIPPED policy's
/// bucket coverage (FLAC hybrid when a profile exists) and carries its bucket's
/// margin. Why: one row per track is everything the report needs, and every dial comes
/// from `truepeak_core::default_policy`.
pub fn evaluate_buckets(
    tracks: &[Track],
    policy: &Policy,
    buckets: &HashMap<String, Bucket>,
    bones: &HashMap<String, Vec<usize>>,
) -> (f64, Vec<BucketRow>) {
    // Fold each track into the decoded total and its measured row.
    let mut decoded: f64 = tracks
        .iter()
        .filter(|track| track.duration_secs <= policy.short_scan_max_secs)
        .map(|track| track.duration_secs)
        .sum();
    let mut rows = Vec::new();
    for track in tracks {
        if track.duration_secs <= policy.short_scan_max_secs {
            continue;
        }
        let bucket = buckets
            .get(&track.path)
            .copied()
            .unwrap_or_else(|| bucket_from_extension(&track.path));
        let track_bones = if bucket == Bucket::Flac { bones.get(&track.path) } else { None };
        let probe = bucket_probe(policy, bucket, track_bones.is_some());
        let even = if track_bones.is_some() {
            policy.bones_even_coverage_fraction
        } else {
            policy.pass1_coverage_fraction.min(probe.coverage_fraction)
        };
        let (peak, used_secs) = hybrid_probe(track, probe.coverage_fraction, even, track_bones);
        decoded += used_secs;
        rows.push(BucketRow {
            full_db: db(f64::from(track.full_peak)),
            probe_db: db(peak),
            margin_db: probe.probe_margin_db,
            bucket,
        });
    }
    (decoded, rows)
}

File: package/music-player/truepeak-core.bench/analysis/buckets.mjs
// Bucket-first policies: join the corpus with the embedded-tag sweep, measure each
// bucket's zoom under-read tail across coverages, then search small per-bucket
// (coverage, margin) tables under the global budget for a composite that beats the
// uniform zoom answer.
import { createReadStream } from 'node:fs';
import { createInterface } from 'node:readline';
import { db, gainErrors, loadTracks, quantile } from './corpus.mjs';

const SHORT_MAX = 90;
const CEILING_DB = -1.0;
const TOO_LOUD = 0.5;
const BUDGET = 229215.18996182326;

const tracks = await loadTracks(new URL('../out/tracks-fine.jsonl', import.meta.url).pathname);
const tags = new Map();
{
  const rl = createInterface({
    input: createReadStream(new URL('../out/tags-full.jsonl', import.meta.url).pathname),
    crlfDelay: Infinity,
  });
  for await (const line of rl) {
    if (!line.trim()) continue;
    const r = JSON.parse(line);
    tags.set(r.path, r);
  }
}

/** Bucket assignment from zero-cost, production-legal observables (no path text). */
function bucketOf(t) {
  const g = tags.get(t.path);
  const ext = t.path.slice(t.path.lastIndexOf('.') + 1).toLowerCase();
  const codec = g?.codec ?? (ext === 'flac' ? 'flac' : ext === 'm4a' || ext === 'mp4' ? 'aac' : ext);
  if (codec === 'flac') return 'flac';
  if (g?.hasStoreIds || g?.hasItunNorm) return 'store';
  if (g?.hasPurl) return 'purl';
  return 'bare';
}

const shorts = tracks.filter((t) => t.dur <= SHORT_MAX);
const longs = tracks.filter((t) => t.dur > SHORT_MAX);
const shortSecs = shorts.reduce((s, t) => s + t.dur, 0);

function makeHeap(bins) {
  const h = [];
  const less = (a, b) => bins[a] < bins[b];
  const push = (i) => {
    h.push(i);
    let c = h.length - 1;
    while (c > 0) {
      const p = (c - 1) >> 1;
      if (less(h[p], h[c])) {
        [h[p], h[c]] = [h[c], h[p]];
        c = p;
      } else break;
    }
  };
  const pop = () => {
    const t = h[0];
    const l = h.pop();
    if (h.length > 0) {
      h[0] = l;
      let p = 0;
      for (;;) {
        const a = 2 * p + 1;
        const b = a + 1;
        let m = p;
        if (a < h.length && less(h[m], h[a])) m = a;
        if (b < h.length && less(h[m], h[b])) m = b;
        if (m === p) break;
        [h[p], h[m]] = [h[m], h[p]];
        p = m;
      }
    }
    return t;
  };
  return { push, pop, size: () => h.length };
}

function zoom(t, cTotal) {
  const bins = t.bins;
  const n = bins.length;
  const budgetBins = Math.max(1, Math.floor(cTotal * n));
  const count = Math.max(1, Math.round(0.1 * n));
  const span = n - 1;
  const decoded = new Uint8Array(n);
  const heap = makeHeap(bins);
  let used = 0;
  let peak = 0;
  const decode = (i) => {
    decoded[i] = 1;
    used += 1;
    if (bins[i] > peak) peak = bins[i];
    heap.push(i);
  };
  for (let k = 0; k < count && used < budgetBins; k += 1) {
    const i = count <= 1 ? Math.floor(span / 2) : Math.round((k / (count - 1)) * span);
    if (!decoded[i]) decode(i);
  }
  while (used < budgetBins && heap.size() > 0) {
    const i = heap.pop();
    if (i > 0 && !decoded[i - 1] && used < budgetBins) decode(i - 1);
    if (i + 1 < n && !decoded[i + 1] && used < budgetBins) decode(i + 1);
  }
  return { peak, used };
}

// Precompute per-track probe results at each candidate coverage (coarse grid).
const COVERAGES = [0.1, 0.14, 0.18, 0.2398, 0.28, 0.32];
const byBucket = new Map();
for (const t of longs) {
  const b = bucketOf(t);
  if (!byBucket.has(b)) byBucket.set(b, []);
  byBucket.get(b).push(t);
}
console.log('long-track buckets (count, seconds, loud count):');
for (const [b, list] of byBucket) {
  const secs = list.reduce((s, t) => s + t.dur, 0);
  const loud = list.filter((t) => db(t.full) > CEILING_DB).length;
  console.log(`  ${b}: n=${list.length} secs=${secs.toFixed(0)} loud=${loud}`);
}

const probes = new Map();
for (const t of longs) {
  const per = new Map();
  for (const c of COVERAGES) per.set(c, db(zoom(t, c).peak));
  probes.set(t.path, per);
}

console.log('\nper-bucket zoom tails by coverage (loud tracks): p90/p99/max under-read; clamps at m=0.4/0.5');
for (const [b, list] of byBucket) {
  const loud = list.filter((t) => db(t.full) > CEILING_DB);
  for (const c of COVERAGES) {
    const urs = loud.map((t) => db(t.full) - probes.get(t.path).get(c));
    const clamps4 = urs.filter((u) => u - 0.4 > TOO_LOUD).length;
    const clamps5 = urs.filter((u) => u - 0.5 > TOO_LOUD).length;
    console.log(
      `  ${b} c=${c}: p90=${quantile(urs, 0.9).toFixed(2)} p99=${quantile(urs, 0.99).toFixed(2)} max=${quantile(urs, 1.0).toFixed(2)} | clamps m0.4=${clamps4} m0.5=${clamps5}`,
    );
  }
}

// Composite search: assign each bucket a coverage and a margin from small grids, subject
// to the global budget; keep worst-quiet at 0.5 (margins <= 0.5) and minimize clamps
// first, then average quiet. Buckets: flac, store, purl, bare.
const MARGINS = [0.3, 0.4, 0.5];
const bucketNames = [...byBucket.keys()];
const bucketSecs = new Map(bucketNames.map((b) => [b, byBucket.get(b).reduce((s, t) => s + t.dur, 0)]));

/** Evaluate one assignment {bucket: {c, m}}; returns measures + decoded seconds. */
function evaluateAssignment(assign) {
  let decoded = shortSecs;
  let clamps = 0;
  let quietSum = 0;
  let worstQuiet = 0;
  for (const [b, list] of byBucket) {
    const { c, m } = assign[b];
    for (const t of list) {
      decoded += Math.max(1, Math.floor(c * t.bins.length)) * t.binSecs;
      const probeDb = probes.get(t.path).get(c);
      const fullDb = db(t.full);
      if (fullDb > CEILING_DB && fullDb - probeDb - m > TOO_LOUD) clamps += 1;
      const { quietDb } = gainErrors({ fullDb, probeDb, marginDb: m, ceilingDb: CEILING_DB });
      quietSum += quietDb;
      worstQuiet = Math.max(worstQuiet, quietDb);
    }
  }
  return { decoded, clamps, avg: quietSum / tracks.length, worstQuiet };
}

// Enumerate: flac and store/purl take low coverages and margins; bare takes what's left.
const results = [];
for (const cFlac of COVERAGES) {
  for (const cStore of COVERAGES) {
    for (const cPurl of COVERAGES) {
      // Bare bucket coverage: the largest grid value that fits the remaining budget.
      const usedFixed = shortSecs +
        cFlac * bucketSecs.get('flac') + cStore * (bucketSecs.get('store') ?? 0) + cPurl * (bucketSecs.get('purl') ?? 0);
      const cBareExact = (BUDGET - usedFixed) / bucketSecs.get('bare');
      const cBare = [...COVERAGES].reverse().find((c) => c <= cBareExact);
      if (!cBare || cBare < 0.14) continue;
      for (const mFlac of MARGINS) {
        for (const mStore of MARGINS) {
          for (const mPurl of MARGINS) {
            for (const mBare of MARGINS) {
              const assign = {
                flac: { c: cFlac, m: mFlac },
                store: { c: cStore, m: mStore },
                purl: { c: cPurl, m: mPurl },
                bare: { c: cBare, m: mBare },
              };
              const r = evaluateAssignment(assign);
              if (r.decoded > BUDGET) continue;
              results.push({ assign, ...r });
            }
          }
        }
      }
    }
  }
}
results.sort((a, b) => a.clamps - b.clamps || a.avg - b.avg);
console.log(`\ncomposite search: ${results.length} in-budget assignments; best by (clamps, avg):`);
for (const r of results.slice(0, 8)) {
  const a = Object.entries(r.assign).map(([b, { c, m }]) => `${b}=(${c},${m})`).join(' ');
  console.log(`  clamps=${r.clamps} avg=${r.avg.toFixed(3)} worst=${r.worstQuiet.toFixed(2)} decoded=${r.decoded.toFixed(0)} | ${a}`);
}
// Also the best assignments that hold average quiet under the uniform answer's 0.371.
const better = results.filter((r) => r.avg <= 0.371 + 1e-9);
better.sort((a, b) => a.clamps - b.clamps || a.avg - b.avg);
console.log('best with avg <= 0.371 (uniform zoom m=0.5):');
for (const r of better.slice(0, 5)) {
  const a = Object.entries(r.assign).map(([b, { c, m }]) => `${b}=(${c},${m})`).join(' ');
  console.log(`  clamps=${r.clamps} avg=${r.avg.toFixed(3)} worst=${r.worstQuiet.toFixed(2)} decoded=${r.decoded.toFixed(0)} | ${a}`);
}
